software/contrib/euclid.py

```diff
@@ -0,0 +1,417 @@
+    @param rot    Optional rotation to offset the pattern. Must be in the range [0, steps]
+    
+    @return An int array of length steps consisting of 1 and 0 values only
+    
+    @exception ValueError if pulses or rot is out of range
+    
+    @license MIT -- https://github.com/brianhouse/bjorklund/blob/master/LICENSE.txt
+    """
+    steps = int(steps)
+    pulses = int(pulses)
+    rot = int(rot)
+    if pulses > steps or pulses < 0:
+        raise ValueError
+    if rot > steps or steps < 0:
+        raise ValueError
+    if steps == 0:
+        return []
+    pattern = []    
+    counts = []
+    remainders = []
+    divisor = steps - pulses
+    remainders.append(pulses)
+    level = 0
+    while True:
+        counts.append(divisor // remainders[level])
+        remainders.append(divisor % remainders[level])
+        divisor = remainders[level]
+        level = level + 1
+        if remainders[level] <= 1:
+            break
+    counts.append(divisor)
+    
+    def build(level):
+        if level == -1:
+            pattern.append(0)
+        elif level == -2:
+            pattern.append(1)         
+        else:
+            for i in range(0, counts[level]):
+                build(level - 1)
+            if remainders[level] != 0:
+                build(level - 2)
+    
+    build(level)
+    i = pattern.index(1)
+    pattern = pattern[i:] + pattern[0:i]
+    
+    # rotate the pattern if needed by removing the last item and
+    # adding it to the start
+    if rot != 0:
+        for i in range(rot):
+            x = pattern.pop(-1)
+            pattern.insert(0, x)
+    
+    return pattern
+
+class EuclidGenerator:
+    """Generates the euclidean rhythm for a single output
+    """
+    
+    def __init__(self, cv_pin, steps=1, pulses=0, rotation=0, skip=0):
+        """Create a generator that sends its output to the given pin
+
+        @param cv_pin  One of the six output jacks (cv1..cv6)
+        @param steps  The initial number of steps (1-32)
+        @param pulses  The initial number of pulses (0-32)
+        @param rotation  The initial rotation (0-32)
+        @param skip  The skip probability (0-1)
+        """
+        
+        ## The CV output this generator controls
+        self.cv = cv_pin
+        
+        ## The current position within the pattern
+        self.position = 0
+        
+        ## The number of steps in the pattern
+        self.steps = steps
+        
+        ## The number of triggers in the pattern
+        self.pulses = pulses
+        
+        ## The rotation of the pattern
+        self.rotation = rotation
+        
+        ## The probability that we skip any given step when it triggers
+        #
+        #  Must be in the range [0, 1], where 0 means never skip and
+        #  1 means always skip
+        self.skip = skip
+        
+        ## The on/off pattern we generate
+        self.pattern = []
+        
+        self.regenerate()
+        
+    def __str__(self):
```

Comment:
[discussion] It looks like each main loop cycle on Channel Menu will regenerate the string pattern each time the draw method is called. I would suggest making the string pattern an instance variable that gets generated at init time, or cache the call to this method: https://docs.python.org/3/library/functools.html#functools.cache

Reply:
Actually, I don't think the cache decorator will work since the __str__ method has no parameters. :(

Reply:
Yes, you're correct in that the string gets re-generated every time `draw()` gets called.  I can add a dirty flag to regenerate it on-change though.  That should lower the CPU usage a bit.

Reply:
Simpler idea: just cache it internally and clear it whenever `advance()` or `regenerate()` gets called.  `__str__` will return the cached copy if it exists, or will just regenerate it.

I'm not familiar with how the `@cache` decorator works and whether or not the fact that the parameter is always `self` will cause some issues with returning the _wrong_ string if it's just a static cache.